flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -266,19 +304,31 @@ protected void restoreJob(
         Optional<String> savepointOpt = Optional.empty();
 
         if (spec.getJob().getUpgradeMode() != UpgradeMode.STATELESS) {
-            savepointOpt =
-                    Optional.ofNullable(
-                                    ctx.getResource()
-                                            .getStatus()
-                                            .getJobStatus()
-                                            .getSavepointInfo()
-                                            .getLastSavepoint())
-                            .flatMap(s -> Optional.ofNullable(s.getLocation()));
+            if (FlinkStateSnapshotUtils.shouldCreateSnapshotResource(
+                    ctx.getOperatorConfig(), deployConfig)) {
+                savepointOpt = getLatestSavepointPathFromFlinkStateSnapshots(ctx);
+            } else {
+                savepointOpt =
+                        Optional.ofNullable(
+                                        ctx.getResource()
+                                                .getStatus()
+                                                .getJobStatus()
+                                                .getSavepointInfo()
+                                                .getLastSavepoint())
+                                .flatMap(s -> Optional.ofNullable(s.getLocation()));
```

Comment:
The problem with this logic is that the snapshotReference field allows us to specify the path directly. So if the user starts the job using that, in the code we don't update the lastSavepointInfo but we also won't find that path here. So we will simply restart from empty state in those cases which is not what should happen.

2 main ways to solve it: 
 1.  We keep the current lastSavepoint logic in the status and update it whenever a snapshot is completed. This is a bit problematic because we would have to update the status of the FlinkDeployment/SessionJob from the snapshot controller, which is a bit weird.
 2. Introduce a new field in the status to capture the savepoint reference used during a Savepoint deployment. I think this is the way to go because this is the functionality that is important to capture. When we submit with initialSavepointPath the first time / savepoint redeploy / savepoint upgrade, in the lastReconciledSpec we have upgradeMode: SAVEPOINT recorded and it is assumed that the lastSavepoint contains the savepoint used. Now we are getting rid of the lastSavepoint so we should put something more meaningful in place in the status to capture the actual upgrade related behaviour. One possibility here is simply to use the `job.snapshotReference` field in the lastReconciledSpec to capture the upgrade savepoint and clear it in other cases (same way of how we reuse the upgradeMode field to capture how we actually stopped the job in the lastReconciledSpec)

Reply:
Actually I think the cleanest / simplest thing to do would be to add a new status field like: 
```
snapshotReference / upgradeSnapshotReference 
```
This should be updated during SAVEPOINT deployments (when we start from savepoint/checkpoint), after savepoint cancellation or when observing the last state of a terminal job. (basically every time we currently update the lastSavepoint for upgrade purposes
)

Reply:
Thank you for detailing this problem so well, it makes perfect sense, I have overlooked this part of the operator.

I think I could tackle this problem with a new status field as you have proposed called `upgradeSnapshotReference` in the job status, but this still leaves an important feature out which was covered before: Manual savepoints triggered by creating a new `FlinkStateSnapshot` resource, and periodic savepoints that create new `FlinkStateSnapshot` resources will not update this new field `upgradeSnapshotReference` in FlinkDeployment/FlinkSessionJob statuses.

I see 3 solutions:
- Update FlinkDeployment/FlinkSessionJob status from the snapshot controller, I agree that this is not an optimal pattern, and I would avoid it.
- List all `FlinkStateSnapshot` resources when re-deploying a Flink job and we check `upgradeSnapshotReference`. We could compare the time the snapshots were taken and use the most recent. This might lead to confusion because e.g. starting a suspended job might not use the savepoint we can see in the status.
- From the FlinkDeployment/FlinkSessionJob controllers check the status of `FlinkStateSnapshot` resources with the help of `InformerEventSource`, not sure how viable this is.

Reply:
@mateczagany I think the only good solution is to introduce a separate `upgradeSnapshotReference` that is only updated during the upgrade cycle (and during terminal job observe).  Any snapshot observed this way is logically speaking always the latest. 

A manual / periodic snapshot should never override this, doing that would be surely an error in the logic so this is absolutely not a problem. This way we will actually separate the upgrade snapshot handling from the savepoint taking / management which was kind of confused together until now.

In other words we don't need any of the "3 solutions" :)

Reply:
Okay, I am definitely happy with this plan, I will push a commit soon:
- the new field will be updated on upgrade, stop-with-savepoint and when the job becomes terminal
- the field won't be set during manual and periodic snapshots (in contrast to how it worked so far)
- savepoint path in `AbstractJobReconciler#restoreJob()` will be set to this field
- `lastSavepoint` will still be used for periodic and manual savepoints if `FlinkStateSnapshot` is disabled. This will be deprecated.

Reply:
I have added a commit that should cover these points, I have also resolved related conversations. The commit also adds support for `savepointTriggerNonce` and `checkpointTriggerNonce` with the new snapshot resources. I will also add unit tests for these later.